File: implementations/typescript/src/relay.ts
/** Tiny ┬º28 HTTP relay. Uses TypeScript validation. Independent of implementations/python/relay.py. */

import { createServer, type IncomingMessage, type ServerResponse } from "node:http";
import { jcsBytes } from "./canon.ts";
import { Fail, Unsupported, validate } from "./check.ts";
import { scanObject } from "./scan.ts";
import type { JsonVal } from "./scan.ts";

const MAXH = (1n << 256n) - 1n;
const multiplier = Math.max(1, Number(process.env.SPP_POW_MULTIPLIER ?? "1"));
const maxRecord = Number(process.env.SPP_MAX_RECORD_BYTES ?? "65536");
const pageSize = Math.max(1, Number(process.env.SPP_PAGE_SIZE ?? "2"));
const blockActor = process.env.SPP_BLOCK_ACTOR ?? "";
const blockChannel = process.env.SPP_BLOCK_CHANNEL ?? "";

type Env = { [k: string]: JsonVal };
const envelopes = new Map<string, Env>();
const advertised: string[] = [];
const channelOf = new Map<string, string[]>();
const objectOf = new Map<string, string[]>();

function push(map: Map<string, string[]>, key: string, id: string): void {
  const arr = map.get(key) ?? [];
  arr.push(id);
  map.set(key, arr);
}

function store(env: Env, ident: string): void {
  const hex = ident.slice(7);
  if (envelopes.has(hex)) return;
  envelopes.set(hex, env);
  const t = env.type;
  if (t === "channel") {
    advertised.push(hex);
    push(channelOf, hex, hex);
    const desc = env.descriptor;
    if (desc && typeof desc === "object" && !Array.isArray(desc) && typeof desc.hash === "string") {
      push(objectOf, desc.hash.slice(7), hex);
    }
  } else if (t === "pointer" && typeof env.channel === "string") {
    push(channelOf, env.channel.slice(7), hex);
    const ref = env.ref;
    if (ref && typeof ref === "object" && !Array.isArray(ref) && typeof ref.hash === "string") {
      push(objectOf, ref.hash.slice(7), hex);
    }
  }
}

function b64url(n: number): string {
  return Buffer.from(String(n), "utf8").toString("base64url");
}

function cursorOk(c: string | null): boolean {
  if (!c) return true;
  return /^[A-Za-z0-9_-]+$/.test(c);
}

function fromCursor(c: string | null): number {
  if (!c) return 0;
  const s = Buffer.from(c, "base64url").toString("utf8");
  const n = Number(s);
  return Number.isFinite(n) && n >= 0 ? n : 0;
}

function slicePage(ids: string[], cursor: string | null): { items: string[]; next: string | null } {
  const start = fromCursor(cursor);
  const chunk = ids.slice(start, start + pageSize);
  const end = start + chunk.length;
  return { items: chunk, next: end < ids.length ? b64url(end) : null };
}

function readBody(req: IncomingMessage): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    let n = 0;
    req.on("data", (c: Buffer) => {
      n += c.length;
      if (n > maxRecord) {
        reject(Object.assign(new Error("too_large"), { code: "too_large" }));
        req.destroy();
        return;
      }
      chunks.push(c);
    });
    req.on("end", () => resolve(Buffer.concat(chunks)));
    req.on("error", reject);
  });
}

function send(res: ServerResponse, code: number, body: unknown): void {
  const data = Buffer.from(JSON.stringify(body));
  res.writeHead(code, { "Content-Type": "application/json", "Content-Length": data.length });
  res.end(data);
}

function err(res: ServerResponse, code: number, error: string, message: string): void {
  send(res, code, { error, message });
}

function workOk(ident: string, units: number): boolean {
  const h = BigInt("0x" + ident.slice(7));
  return h <= MAXH / (BigInt(units) * 65536n * BigInt(multiplier));
}

const server = createServer(async (req, res) => {
  const url = new URL(req.url ?? "/", "http://127.0.0.1");
  const path = url.pathname;
  const cursor = url.searchParams.get("cursor");
  try {
    if (req.method === "GET" && path === "/.well-known/spp") {
      return send(res, 200, {
        v: 1,
        submit: "/v1/assertions",
        channels: "/v1/channels",
        assertion: "/v1/assertions/sha256/{hex}",
        channel: "/v1/channels/sha256/{hex}/assertions",
        object: "/v1/objects/sha256/{hex}/assertions",
        bootstrap_channels: [],
        policy: { pow_multiplier: multiplier, max_record_bytes: maxRecord },
      });
    }
    if (req.method === "GET" && path === "/v1/channels") {
      if (!cursorOk(cursor)) return err(res, 400, "bad_cursor", "cursor must match the base64url alphabet");
      const ids = advertised.map((h) => `sha256:${h}`);
      const page = slicePage(ids, cursor);
      return send(res, 200, { items: page.items, next: page.next });
    }
    let m = /^\/v1\/assertions\/sha256\/([0-9a-f]{64})$/.exec(path);
    if (req.method === "GET" && m) {
      const env = envelopes.get(m[1]!);
      if (!env) return err(res, 404, "not_found", "unknown assertion");
      return send(res, 200, JSON.parse(Buffer.from(jcsBytes(env)).toString("utf8")));
    }
    m = /^\/v1\/channels\/sha256\/([0-9a-f]{64})\/assertions$/.exec(path);
    if (req.method === "GET" && m) {
      if (!cursorOk(cursor)) return err(res, 400, "bad_cursor", "cursor must match the base64url alphabet");
      const page = slicePage(channelOf.get(m[1]!) ?? [], cursor);
      return send(res, 200, { items: page.items.map((h) => envelopes.get(h)), next: page.next });
    }
    m = /^\/v1\/objects\/sha256\/([0-9a-f]{64})\/assertions$/.exec(path);
    if (req.method === "GET" && m) {
      if (!cursorOk(cursor)) return err(res, 400, "bad_cursor", "cursor must match the base64url alphabet");
      const page = slicePage(objectOf.get(m[1]!) ?? [], cursor);
      return send(res, 200, { items: page.items.map((h) => envelopes.get(h)), next: page.next });
    }
    if (req.method === "POST" && path === "/v1/assertions") {
      const enc = (req.headers["content-encoding"] ?? "").toString();
      if (enc && enc.toLowerCase() !== "identity") return err(res, 400, "bad_encoding", "non-identity Content-Encoding");
      const ctype = (req.headers["content-type"] ?? "").toString().split(";")[0]!.trim();
      if (ctype !== "application/json" && ctype !== "application/spp+json") {
        return err(res, 400, "bad_type", "Content-Type");
      }
      const cl = Number(req.headers["content-length"] ?? "0");
      if (cl > maxRecord) return err(res, 400, "too_large", "max_record_bytes");
      const body = await readBody(req);
      if (body.length > maxRecord) return err(res, 400, "too_large", "max_record_bytes");
      let info: { id: string; units: number };
      try {
        info = validate(new Uint8Array(body));
      } catch (e) {
        if (e instanceof Unsupported) return err(res, 400, "unsupported_assertion", e.reason);
        if (e instanceof Fail) return err(res, 400, "invalid_assertion", `${e.stage}:${e.reason}`);
        throw e;
      }
      const env = scanObject(new Uint8Array(body)) as Env;
      if (blockActor && env.actor === blockActor) return err(res, 403, "policy", "actor blocked");
      if (blockChannel && env.channel === blockChannel) return err(res, 403, "policy", "channel blocked");
      if (!workOk(info.id, info.units)) return err(res, 403, "policy", "local proof-of-work multiplier");
      const existed = envelopes.has(info.id.slice(7));
      store(env, info.id);
      return send(res, 200, { id: info.id, duplicate: existed });
    }
    return err(res, 404, "not_found", "route");
  } catch (e) {
    if ((e as { code?: string }).code === "too_large") return err(res, 400, "too_large", "max_record_bytes");
    console.error(e);
    return err(res, 500, "internal", "error");
  }
});

const port = (() => {
  const i = process.argv.indexOf("--port");
  return i >= 0 ? Number(process.argv[i + 1]) : 18761;
})();

server.listen(port, "127.0.0.1", () => {
  console.log(`relay-b http://127.0.0.1:${port}`);
});

File: implementations/python/relay.py
#!/usr/bin/env python3
"""Tiny §28 HTTP relay. Uses Python reference validation. No shared server with TypeScript."""

from __future__ import annotations

import json
import os
import re
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))

from jcs import canonicalize_bytes  # noqa: E402
from protocol import SppError, UnsupportedError, validate_bytes  # noqa: E402
from json_input import parse_object  # noqa: E402

MAX_H = (1 << 256) - 1
HEX64 = 64
HEX64_RE = re.compile(r"^[0-9a-f]{64}$")


class Store:
    def __init__(self) -> None:
        self.envelopes: dict[str, dict] = {}
        self.order: list[str] = []
        self.channel_of: dict[str, list[str]] = {}
        self.object_of: dict[str, list[str]] = {}
        self.advertised: list[str] = []
        self.multiplier = max(1, int(os.environ.get("SPP_POW_MULTIPLIER", "1")))
        self.max_record = int(os.environ.get("SPP_MAX_RECORD_BYTES", "65536"))
        self.page = max(1, int(os.environ.get("SPP_PAGE_SIZE", "2")))
        self.block_actor = os.environ.get("SPP_BLOCK_ACTOR", "")
        self.block_channel = os.environ.get("SPP_BLOCK_CHANNEL", "")

    def put(self, env: dict, ident: str) -> None:
        digest = ident[7:]
        if digest in self.envelopes:
            return
        self.envelopes[digest] = env
        self.order.append(digest)
        u = {k: v for k, v in env.items() if k not in ("id", "sig")}
        if u.get("type") == "channel":
            self.advertised.append(digest)
            self.channel_of.setdefault(digest, []).append(digest)
            desc = u.get("descriptor") or {}
            h = desc.get("hash")
            if isinstance(h, str) and h.startswith("sha256:"):
                self.object_of.setdefault(h[7:], []).append(digest)
        elif u.get("type") == "pointer":
            ch = str(u.get("channel", ""))
            if ch.startswith("sha256:"):
                self.channel_of.setdefault(ch[7:], []).append(digest)
            ref = u.get("ref") or {}
            h = ref.get("hash")
            if isinstance(h, str) and h.startswith("sha256:"):
                self.object_of.setdefault(h[7:], []).append(digest)


STORE = Store()


def page(items: list[str], cursor: str | None) -> tuple[list[str], str | None]:
    try:
        start = int(cursor) if cursor else 0
    except ValueError:
        raise ValueError("bad-cursor")  # handled by the request layer
    start = max(0, start)
    chunk = items[start : start + STORE.page]
    nxt = start + len(chunk)
    return chunk, (str(nxt) if nxt < len(items) else None)


def local_work_ok(ident: str, units: int) -> bool:
    h = int(ident[7:], 16)
    return h <= MAX_H // (units * 65536 * STORE.multiplier)


class Handler(BaseHTTPRequestHandler):
    def log_message(self, fmt: str, *args) -> None:
        sys.stderr.write("relay-a: " + fmt % args + "\n")

    def _send(self, code: int, body, ctype: str = "application/json") -> None:
        data = body if isinstance(body, (bytes, bytearray)) else json.dumps(body).encode("utf-8")
        self.send_response(code)
        self.send_header("Content-Type", ctype)
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def _err(self, code: int, token: str, message: str) -> None:
        self._send(code, {"error": token, "message": message})

    def do_GET(self) -> None:
        path = urlparse(self.path).path
        q = parse_qs(urlparse(self.path).query)
        cursor = q.get("cursor", [None])[0]
        if cursor is not None:
            try:
                int(cursor)
            except ValueError:
                return self._err(400, "bad_cursor", "cursor must be an integer")
        if path == "/.well-known/spp":
            return self._send(
                200,
                {
                    "v": 1,
                    "submit": "/v1/assertions",
                    "channels": "/v1/channels",
                    "assertion": "/v1/assertions/sha256/{hex}",
                    "channel": "/v1/channels/sha256/{hex}/assertions",
                    "object": "/v1/objects/sha256/{hex}/assertions",
                    "bootstrap_channels": [],
                    "policy": {
                        "pow_multiplier": STORE.multiplier,
                        "max_record_bytes": STORE.max_record,
                    },
                },
            )
        if path == "/v1/channels":
            ids, nxt = page([f"sha256:{h}" for h in STORE.advertised], cursor)
            return self._send(200, {"items": ids, "next": nxt})
        parts = path.strip("/").split("/")
        if parts[:3] == ["v1", "assertions", "sha256"] and len(parts) == 4:
            if not HEX64_RE.fullmatch(parts[3]):
                return self._err(404, "not_found", "route")
            env = STORE.envelopes.get(parts[3])
            if not env:
                return self._err(404, "not_found", "unknown assertion")
            return self._send(200, canonicalize_bytes(env))
        if parts[:3] == ["v1", "channels", "sha256"] and len(parts) == 5 and parts[4] == "assertions":
            if not HEX64_RE.fullmatch(parts[3]):
                return self._err(404, "not_found", "route")
            ids, nxt = page(STORE.channel_of.get(parts[3], []), cursor)
            items = [STORE.envelopes[i] for i in ids]
            return self._send(200, {"items": items, "next": nxt})
        if parts[:3] == ["v1", "objects", "sha256"] and len(parts) == 5 and parts[4] == "assertions":
            if not HEX64_RE.fullmatch(parts[3]):
                return self._err(404, "not_found", "route")
            ids, nxt = page(STORE.object_of.get(parts[3], []), cursor)
            items = [STORE.envelopes[i] for i in ids]
            return self._send(200, {"items": items, "next": nxt})
        return self._err(404, "not_found", "route")

    def do_POST(self) -> None:
        path = urlparse(self.path).path
        if path != "/v1/assertions":
            return self._err(404, "not_found", "route")
        enc = self.headers.get("Content-Encoding", "")
        if enc and enc.lower() not in ("identity",):
            return self._err(400, "bad_encoding", "non-identity Content-Encoding")
        ctype = (self.headers.get("Content-Type") or "").split(";")[0].strip()
        if ctype not in ("application/json", "application/spp+json"):
            return self._err(400, "bad_type", "Content-Type")
        n = int(self.headers.get("Content-Length") or "0")
        if n > STORE.max_record:
            return self._err(400, "too_large", "max_record_bytes")
        body = self.rfile.read(n)
        if len(body) > STORE.max_record:
            return self._err(400, "too_large", "max_record_bytes")
        try:
            info = validate_bytes(body)
        except UnsupportedError as e:
            return self._err(400, "unsupported_assertion", e.reason)
        except SppError as e:
            return self._err(400, "invalid_assertion", f"{e.stage}:{e.reason}")
        env = parse_object(body)
        actor = str(env.get("actor", ""))
        channel = str(env.get("channel", ""))
        if STORE.block_actor and actor == STORE.block_actor:
            return self._err(403, "policy", "actor blocked")
        if STORE.block_channel and channel == STORE.block_channel:
            return self._err(403, "policy", "channel blocked")
        if not local_work_ok(info["id"], info["units"]):
            return self._err(403, "policy", "local proof-of-work multiplier")
        existed = info["id"][7:] in STORE.envelopes
        STORE.put(env, info["id"])
        self._send(200, {"id": info["id"], "duplicate": existed})


def main() -> None:
    port = 18760
    args = sys.argv[1:]
    if "--port" in args:
        port = int(args[args.index("--port") + 1])
    httpd = ThreadingHTTPServer(("127.0.0.1", port), Handler)
    print(f"relay-a http://127.0.0.1:{port}", flush=True)
    httpd.serve_forever()


if __name__ == "__main__":
    main()
